member_map.go

```diff
@@ -17,66 +17,215 @@
+
+	// Ip address
 	Addr net.IP
 
-	// port
+	// Port
 	Port uint16
 
 	// Current member status from my point of view
-	status Status
+	Status Status
 
 	// Time last status change happened
-	lastStatusChange time.Time
+	LastStatusChange time.Time
 
 	// Incarnation helps to keep the most fresh information about member status in the system
 	// which tells that suspect member confirming that it is alive, and only when suspect
 	// got suspicion message, that member can increments incarnation
-	incarnation uint32
+	Incarnation uint32
 }
 
-// convert member addr and port to string
+// Convert member addr and port to string
 func (m *Member) Address() string {
 	return net.JoinHostPort(m.Addr.String(), strconv.Itoa(int(m.Port)))
 }
 
+// Get
+func (m *Member) GetID() MemberID {
+	return m.ID
+}
+
 type MemberMap struct {
-	sync.RWMutex
-	members map[MemberID]*Member
+	lock    sync.RWMutex
+	members map[MemberID]Member
+
+	// This is for selecting k random member based on round-robin
+	waitingMembers []MemberID
+}
+
+func NewMemberMap() *MemberMap {
+	return &MemberMap{
+		members:        make(map[MemberID]Member),
+		waitingMembers: make([]MemberID, 0),
+		lock:           sync.RWMutex{},
+	}
 }
 
-// select K random member from members
-func (m *MemberMap) selectKRandomMember(k int) []Member {
+// Select K random memberID from waitingMembers(length of returning member can be lower than k).
+// ** WaitingMembers are shuffled every time when members are updated **, so just returning first K item in waitingMembers is same as
+// selecting k random membersID.
+func (m *MemberMap) SelectKRandomMemberID(k int) []MemberID {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	// If length of members is lower then k,
+	// return current waitingMembers and reset waitingMembers.
+	if len(m.waitingMembers) < k {
+		kMembers := m.waitingMembers
+		defer func() { m.waitingMembers = resetWaitingMembersID(m.members) }()
+		return kMembers
+	}
+
+	// Remove first k membersID.
+	kMembers := m.waitingMembers[:k]
+	m.waitingMembers = m.waitingMembers[k:len(m.waitingMembers)]
+
+	if len(m.waitingMembers) == 0 {
+		m.waitingMembers = resetWaitingMembersID(m.members)
+	}
+
+	return kMembers
+}
+
+func (m *MemberMap) GetMembers() []Member {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	members := make([]Member, 0)
+	for _, v := range m.members {
+		members = append(members, v)
+	}
+
+	return members
+}
+
+func (m *MemberMap) AddMember(member Member) error {
+	m.lock.Lock()
+	defer m.lock.Unlock()
+
+	// If id is empty return error
+	if member.GetID().ID == "" {
+		return ErrEmptyMemberID
+	}
+
+	// Check whether it is already exist
+	exMem, ok := m.members[member.GetID()]
+	if ok {
+		// Apply override Rule
+		member = override(member, exMem)
+	}
+
+	m.members[member.GetID()] = member
+
+	m.waitingMembers = resetWaitingMembersID(m.members)
+	return nil
+}
+
+// Update member and update waitingMembers
+// todo
+func (m *MemberMap) UpdateMember(member Member) error {
+	return nil
+}
```

Comment:
I think it is very similar to `AddMember`. What do you think?

Reply:
I agree. I should think about all cases, if `add and update function` do same things, then update and add function will be merged. So i marked this function(todo)